### PR TITLE
Adds helpers with known middle layer state, used to create resources based on schema alone.

### PR DESCRIPTION
**Pull request by thomasfinstad**
Some cleanup and testing will be needed, I have only done very peliminary testing myself so far, but will do more over time.
The goal is the have a known middle layer/state/structure that can be created/transformed to/from vyos api and terraform, resulting in a easy and quick way to add resources by just designing the schema and some extra bits of "meta data" for the resource.

The code is probably awefully slow compared to what it can be, and the logging needs an overhaul, but I would like to see what you think of the consept before I do much more.

PS [the pull request for vyos-client](https://github.com/Foltik/vyos-client-go/pull/6) is needed for this to work.

The firewall rule and dhcp server is the current test resorces. I was thinking of making the subnet part of the dhcp server its own resource to test out how that would work.

### CODE REVIEW

**Comment by foltik** [comment]
Just merged the needed functionality to vyos-client-go, will look over this soon!

**Comment by thomasfinstad** [comment]
> Just merged the needed functionality to vyos-client-go, will look over this soon!

Please let me know if you have any questions etc, as I know it is a bit of a code dump, and not all that readable.

**One more note:** The resources that rely on these helpers must be a _mirror_ of the vyos config structure.

**Comment by thomasfinstad** [comment]
I have been busy adding some more resources, and fixing bugs. I hope to soon be able to get my entire vyos config into terraform once I have enough resources supported so I can properly dog-food it as compansation for our current lack of tests.

Once I have that done I will request a review.

**Comment by benjamin-pilgrim** [comment]
Hey @thomasfinstad, I tried to add support for the tftp-server service using your fork as a base, and I'm running into issues because of [`service tftp-server listen-address`](https://docs.vyos.io/en/stable/configuration/service/tftp-server.html#cfgcmd-set-service-tftp-server-listen-address-address).

I can't create a child resource because the listen-address must be comitted at the same time as the tftp-service to be valid.

Any thoughts around how this should be handled? I wonder if it's possible to have a KeyTemplate on a Resource within a TypeSet or something.

**Comment by thomasfinstad** [comment]
@TGNThump After a bit of work I have gotten the service tftp-server to work for me, however I have done very minimal testing, and there are some quirks, but you can check out the new commit and give the new resource a try.

**Comment by benjamin-pilgrim** [comment]
> @TGNThump After a bit of work I have gotten the service tftp-server to work for me, however I have done very minimal testing, and there are some quirks, but you can check out the new commit and give the new resource a try.

Had a brief look through the diff. Haven't had a chance to try it out yet, but I see that the way you added support for listen_address wouldn't support the vrf option, which is probably not ideal if other places in the vyos config follow the same pattern.

My thoughts were along the lines of supporting something like this, hijacking the description field of the nested resource:

```go
Schema: map[string]*schema.Schema{
      "listener": {
	      Description: "Listen address.",
	      Type:        schema.TypeSet,
	      Elem: &schema.Resource{
		      Description: "listen-address {{address}}",
		      Schema: map[string]*schema.Schema{
			      "address": {
				      Description: "address",
				      Type:        schema.TypeString,
				      Required:    true,
			      },
			      "vrf": {
				      Description: "vrf",
				      Type:        schema.TypeString,
				      Optional:    true,
			      },
		      },
	      },
	      Required: true,
      },
},
```
with terraform
```terraform
resource "vyos_tftp_service" "default" {
  directory = "/config/tftpboot"

  listener {
    address = "10.0.0.1"
    vrf = "test"
  }
}
```
mapping to
```
service {
  tftp-server {
    listen-address 10.0.0.1 {
      vrf test
    }
  }
}
```

Thoughts?

File: vyos/provider.go [hunk 1]
```diff
@@ -49,6 +49,15 @@ func Provider() *schema.Provider {
 			"vyos_dhcp_server_subnet":              resourceInfoDhcpServerSubnet().ResourceSchema,
 			"vyos_dhcp_server_subnet_address_pool": resourceInfoDhcpServerSubnetAddressPool().ResourceSchema,
 			"vyos_vrrp_group":                      resourceInfoVrrpGroup().ResourceSchema,
+			"vyos_interface_ethernet":              resourceInterfaceEthernet().ResourceSchema,
+			"vyos_interface_ethernet_vif":          resourceInterfaceEthernetVif().ResourceSchema,
+			"vyos_interface_pseudo_ethernet":       resourceInterfacePseudoEthernet().ResourceSchema,
+			"vyos_intercafe_pseudo_ethernet_vif":   resourceInterfacePseudoEthernetVif().ResourceSchema,
```
Comment (benjamin-pilgrim) [inline vyos/provider.go]: Misspelling of interface in resource name

Comment (thomasfinstad) [reply vyos/provider.go]: ty

**Comment by thomasfinstad** [comment]
> My thoughts were along the lines of supporting something like this, hijacking the description field of the nested resource:

I like the idea of being able to extend my crazy little framework, however doing it in the description would leak into the documentation, I think that will create confusion for many.

I noticed vrf in the documentation, but it did not show up as an option on my own box so I had to skip it.

What I have done earlier in these types of situations is create a seperate resource for it, see dhcp server + subnet. Though this is not an ideal solution at all, especially since terraform can't do api batching making terraform for vyos fairly slow, it is currently the best I can do with this hacky framework I made. A seperate resource might be the best solution for now until we are ready to move on to the next chapter with something better than this.

This is also my first walk into golang, and I learned a bit, and in the future I think it would be much wiser to look into either vyos new [graphql](https://phabricator.vyos.net/T3993) api or use some type of api object inspection and auto generate the terraform schemas by just guiding it a little instead of manual recreation of the same structure of the api in terraform schemas: https://blog.vyos.io/writing-vyos-operational-mode-commands-just-got-better

**Comment by benjamin-pilgrim** [comment]
@thomasfinstad I'm seeing the following error:
```
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:61] key_string: 'listen_address' parent_vyos_native_config: 'map[string]interface {}{\"directory\":\"/config/tftpboot\", \"listen-address\":map[string]interface {}{\"10.0.0.1\":map[string]interface {}{}}}'","@timestamp":"2022-1
0-25T17:34:25.010055+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [structures.go:GetChild:184] {\u0026{service tftp-server}} Fetching child: listen_address","@timestamp":"2022-10-25T17:34:25.010093+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [structures.go:CreateChild:140] {\u0026{service tftp-server}} Creating child: listen_address","@timestamp":"2022-10-25T17:34:25.010132+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:45] {listen_address} Walking","@timestamp":"2022-10-25T17:34:25.010168+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:80] resource_schema.Type: TypeList","@timestamp":"2022-10-25T17:34:25.010208+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:81] parent_vyos_native_config: map[string]interface {}{\"10.0.0.1\":map[string]interface {}{}}","@timestamp":"2022-10-25T17:34:25.010248+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:111] resource_schema set/list","@timestamp":"2022-10-25T17:34:25.010282+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:145] resource_schema_elem: '*schema.Schema{Type:4, ConfigMode:0, Required:false, Optional:false, Computed:false, ForceNew:false, DiffSuppressFunc:(schema.SchemaDiffSuppressFunc)(nil), DiffSuppressOnRefresh:false, Default:interf
ace {}(nil), DefaultFunc:(schema.SchemaDefaultFunc)(nil), Description:\"\", InputDefault:\"\", StateFunc:(schema.SchemaStateFunc)(nil), Elem:interface {}(nil), MaxItems:0, MinItems:0, Set:(schema.SchemaSetFunc)(nil), ComputedWhen:[]string(nil), ConflictsWith:[]string(nil), ExactlyOneOf:[]string(nil), AtLeast
OneOf:[]string(nil), RequiredWith:[]string(nil), Deprecated:\"\", ValidateFunc:(schema.SchemaValidateFunc)(nil), ValidateDiagFunc:(schema.SchemaValidateDiagFunc)(nil), Sensitive:false}'","@timestamp":"2022-10-25T17:34:25.010338+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [TRACE] [factories.go:vyosWalker:149] adding value: parent_vyos_native_config: 'map[string]interface {}{\"10.0.0.1\":map[string]interface {}{}}'","@timestamp":"2022-10-25T17:34:25.010383+01:00"}
{"@level":"info","@message":"2022/10/25 17:34:25 [ERROR] [factories.go:vyosWalker:157] resource_schema_elem.Type is not string or list of string: 4","@timestamp":"2022-10-25T17:34:25.010419+01:00"}
```
with config:
```
resource "vyos_tftp_service" "default" {
  directory = "/config/tftpboot"

  listen_address = ["10.0.0.1"]
}
```

**Comment by benjamin-pilgrim** [comment]
@thomasfinstad
> I like the idea of being able to extend my crazy little framework, however doing it in the description would leak into the documentation, I think that will create confusion for many.

Makes sense, I wonder if we could do something sneaky like using type embedding Resource into our own struct.

> I noticed vrf in the documentation, but it did not show up as an option on my own box so I had to skip it.

I think `vrf` was added after the last LTS version.

> What I have done earlier in these types of situations is create a seperate resource for it, see dhcp server + subnet. Though this is not an ideal solution at all, especially since terraform can't do api batching making terraform for vyos fairly slow, it is currently the best I can do with this hacky framework I made. A seperate resource might be the best solution for now until we are ready to move on to the next chapter with something better than this.

I don't think this would work with listen-address because we have to commit listen-address and the parent tftp-server block at the same time for it to be valid.

> This is also my first walk into golang, and I learned a bit, and in the future I think it would be much wiser to look into either vyos new [graphql](https://phabricator.vyos.net/T3993) api or use some type of api object inspection and auto generate the terraform schemas by just guiding it a little instead of manual recreation of the same structure of the api in terraform schemas: https://blog.vyos.io/writing-vyos-operational-mode-commands-just-got-better

The source of truth for the schema is the [interface definitions](https://github.com/vyos/vyos-1x/tree/current/interface-definitions). Code generating the resources from those would be pretty cool, although as you say it would require some manual effort to define what config blocks form a terraform resource ect😅

**Comment by thomasfinstad** [comment]
> I'm seeing the following error:

@TGNThump it seems you get a map returned and I get a string/list of string, this might be because you have the vrf option and I do not. I will try to update my boxes to the latest rolling and see if that changes anything for me.

**Comment by thomasfinstad** [comment]
@TGNThump i have looked into it a bit and I think I will remove the tftp resource for now.

I recommend you see if you can do it with the config block type resource, as we would have to do some large changes or additions to make it work well.

I think I will just continue a bit to get my own config up and running for now, and once I can exploit that work to become more efficient I will start a serious look at possibilities of more auto generated stuff using the interfaces.

Sorry to bring sad news, but it would be a bit hard to add compability for it, of course you are free to make a normal stand alone resource if you wish.

**Comment by thomasfinstad** [comment]
Due to continous changes in the rolling release api I will close this and open a new PR when I have something based on something more automatic that can help us keep up with rolling changes.

**Pull request closed without merging**